flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitReadOperator.java

```diff
@@ -0,0 +1,368 @@
+
+		SplitReader(FileInputFormat<OT> format,
+					TypeSerializer<OT> serializer,
+					TimestampedCollector<OT> collector,
+					Object checkpointLock) {
+
+			this.format = checkNotNull(format, "Unspecified FileInputFormat.");
+			this.serializer = checkNotNull(serializer, "Unspecified Serialized.");
+
+			this.pendingSplits = new LinkedList<>();
+			this.collector = collector;
+			this.checkpointLock = checkpointLock;
+			this.isRunning = true;
+		}
+
+		void addSplit(FileInputSplit split) {
+			Preconditions.checkNotNull(split);
+			synchronized (lock) {
+				this.pendingSplits.add(split);
+			}
+		}
+
+		public boolean isRunning() {
+			return this.isRunning;
+		}
+
+		@Override
+		public void run() {
+			FileInputSplit split = null;
+			try {
+				while (this.isRunning) {
+
+					// get the next split to read.
+					// locking is needed because checkpointing will
+					// ask for a consistent snapshot of the list.
+					synchronized (lock) {
+						split = this.pendingSplits.peek();
+					}
+
+					if (split == null) {
+						Thread.sleep(50);
+						continue;
+					}
+
+					if (split.equals(EOF)) {
+						isRunning = false;
+						break;
+					}
+
+					synchronized (checkpointLock) {
+						synchronized (lock) {
+							split = this.pendingSplits.poll();
+						}
+						this.format.open(split);
+					}
+
+					try {
+						OT nextElement = serializer.createInstance();
+						do {
+							synchronized (checkpointLock) {
+								nextElement = format.nextRecord(nextElement);
+								if (nextElement != null) {
+									collector.collect(nextElement);
+								}
+							}
+						} while (nextElement != null && !format.reachedEnd());
+					} finally {
+						this.format.close();
+					}
+				}
+				LOG.info("Split Reader terminated, and exiting normally.");
+
+			} catch (Throwable e) {
+				if (isRunning) {
+					LOG.error("Caught exception processing split: ", split);
+				}
+				getContainingTask().failExternally(e);
+			}
+		}
+
+		Tuple3<List<FileInputSplit>, FileInputSplit, S> getReaderState() throws IOException {
+			// take a consistent snapshot of the pending splits
+			// and that of the format, which includes the split currently being read and
+			// where we are in the split.
+
+			List<FileInputSplit> snapshot;
+			Tuple2<FileInputSplit, S> formatState = null;
+			synchronized (lock) {
+				snapshot = new ArrayList<>(this.pendingSplits.size());
+				for (FileInputSplit split: this.pendingSplits) {
+					snapshot.add(split);
+				}
+
+				if (this.format instanceof CheckpointableInputFormat) {
+					formatState = ((CheckpointableInputFormat) format).getCurrentChannelState();
+				} else {
+					LOG.warn("The format used is not checkpointable.");
+				}
+			}
+
+			return formatState == null ?
+				new Tuple3<>(snapshot, (FileInputSplit) null, (S) null) :
+				new Tuple3<>(snapshot, formatState.f0, formatState.f1);
+		}
+
+		void setReaderState(Tuple3<List<FileInputSplit>, FileInputSplit, S> state) throws IOException {
```

Comment:
Can we have an additional constructor instead of this method. It is only ever called right after the constructor. This would make it "atomic".